### ISSUE TITLE
Allow style definition to include defaults

### ISSUE THREAD

**Issue by nvkelso** [tangrams/tangram#489]
In most the Mapzen basemap styles we create a `style` and then a special `layer` with a draw section that sets up required defaults for that style (and is usually linked up to data `source`). 

The downside is when someone wants to use the defaults in our house basemap styles, they must use exactly the data source / layer names to enjoy the sensible draw style defaults. Or they need to dip into the YAML source code and copy our defaults to a custom layer.

Many other mapping and graphics applications allow defaults to be stored with at the `style` level, I propose Tangram also enable this.

**Comment by nvkelso** [comment]
**Example 1:**

To make use of Mapzen's icon library, a developer would need to restate many of the property settings from the `pois` layer. Critically this includes the sprite's `size`, `sprite` lookup function, and `sprite_default`. 

From Walkabout, but applies to all house styles:

```
styles:
    icons:
        base: points
        texture: pois
        blend_order: 1

layers:
    pois:
        data: { source: mapzen, layer: pois }
        #visible: global.label_visible_poi_landuse
        filter:
            all:
                - not: { kind: [building,address,farm,apron,residential,commercial,industrial] }
                - $geometry: point
                # White list a few kinds to "always" show per the tile zoom inclusion, else test feature's min_zoom
                #- function() { return feature.min_zoom <= $zoom }
                - function() { if( feature.kind == 'attraction' || feature.kind == 'landmark' || feature.kind == 'hunting' || feature.kind == 'dive_centre' || feature.kind == 'gas_canister' || feature.kind == 'hunting' || feature.kind == 'fishing_area' || feature.kind == 'playground' || feature.kind == 'toilets' ) { return true } else { return feature.min_zoom <= $zoom } }
        draw:
            icons:
                visible: global.icon_visible_poi_landuse
                size: [[13, 18px], [16, 18px], [18, 22px]]
                sprite: function() { return feature.kind; }
                sprite_default: generic
                interactive: global.interactive
                priority: 65 #function() { return (feature.min_zoom && Math.floor(feature.min_zoom * 1000)) || 65; }
                repeat_group: abc
                buffer: 7px
                text:
                    visible: global.text_visible_poi_landuse
                    text_source: global.ux_language_text_source
                    text_wrap: 18
                    max_lines: 3
                    font:
                        family: global.text_font_family
                        weight: 600
                        fill: [0.10,0.10,0.10]
                        size: [[13,10px],[14,10px],[18,11px]]
                        stroke: { color: global.text_stroke, width: 3 }
```

**Example 2:**

To make use of Mapzen's icon library road and transit shields, a developer would need to restate many of the property settings from sublayers in the `roads` layer. Critically this includes the sprite's `size`, `sprite` lookup function, and `sprite_default`. 

From Walkabout, but applies to all house styles.

**Comment by nvkelso** [comment]
**Example 3:** 

Dashed lines.

```
styles:
    dashed:
        base: lines
        dash: [1.25, 1.25]
        dash_background_color: [0.808,0.860,0.851]
layers:
  roads:
        steps:
            filter:
                all:
                    - kind: path
                    - kind_detail: steps
            draw:
                dashed:
                    #order: function() { return feature.sort_rank; }
                    order: 1005
                    color: [0.501,0.650,0.650]
                    width: [[15, 0.6], [16, 1px], [17, 1px], [18, 4px], [19, 1.5m]]
```

Versus:

```
styles:
    dashed:
        base: lines
        dash: [1.25, 1.25]
        dash_background_color: [0.808,0.860,0.851]
        color: [0.501,0.650,0.650]
        width: [[15, 0.6], [16, 1px], [17, 1px], [18, 4px], [19, 1.5m]]
layers:
  roads:
        steps:
            filter:
                all:
                    - kind: path
                    - kind_detail: steps
            draw:
                dashed:
                    #order: function() { return feature.sort_rank; }
                    order: 1005
```

**Comment by pnorman** [comment]
I find example 3 the easiest to understand.

If I'm understanding this right, it'd be the same amount to copy/paste, but what you need to copy/paste is all together? The second way seems more logical to me, and promotes reuse within a style too.

**Comment by bcamper** [comment]
@matteblair and I are discussing prototyping this in JS to see how it feels in practice

**Comment by nvkelso** [comment]
Related documentation about `mix`: https://mapzen.com/documentation/tangram/Styles-Overview/#style-composition-with-mix has some shared concepts.

**Comment by nvkelso** [comment]
See confusion around existing setup in https://github.com/tangrams/tangram-es/issues/1273.

**Comment by matteblair** [comment]
That issue is more of a quirk of Markers in tangram-es, not strictly related to this.

**Comment by meetar** [comment]
> The downside is when someone wants to use the defaults in our house
basemap styles

To clarify the problem: should this sentence read "someone wants to
*change* the defaults"? Or is it more like "someone wants to use default
settings without using the whole basemap"?

**Comment by bcamper** [comment]
I have an implementation of this here:

https://github.com/tangrams/tangram/compare/draw-defaults

The syntax is to add an optional `draw` block to the style definition. This is slightly different from the examples above but is more explicit about which parameters are default `draw` parameters, vs. other style-specific parameters.

The following example provides defaults for `size` and `sprite` parameters:

```
styles:
    icons:
        base: points
        texture: icons
        draw:
            size: 20px
            sprite: function() { return feature.kind; }
```

**Comment by bcamper** [comment]
Shall we target this for v0.12.0? It's a simple implementation (11 lines!). Mostly a matter of whether we agree on the need.

**Comment by nvkelso** [comment]
I still dig it :)

**Comment by bcamper** [comment]
@matteblair are you good with this?

**Comment by tallytalwar** [comment]
Looks good to me from ES implementation point of view.

**Comment by matteblair** [comment]
I still like the principle here, I'm in favor of adding it to see how much it helps.

It might take more than 11 lines in tangram-es but we already have node-merging code so it shouldn't be hard to implement.

**Comment by bcamper** [comment]
Released in [v0.12.0](https://github.com/tangrams/tangram/releases/tag/v0.12.0)
